### PR TITLE
Harmonic model return residuals

### LINKED ISSUE #459: Harmonic model testing
Increasing the number of years in the testing of the harmonic model fit in the following test

https://github.com/MESMER-group/mesmer/blob/dc60db6b1e154bde73f5c242a71792b458f41224/tests/unit/test_harmonic_model.py#L206-L241

fails. The selected orders increase even though the fit for the coefficients is close to zero. We should check the BIC and why it keeps fitting in this case. The fitting still works on real data, so I'm really not sure what is happening here. Even if it does not change the predictions significatntly, fitting up to the max order on every gridpoint massively slows down the fit.

Things we should check: 
- the BIC
- if it maybe works when the trend in the global temperature is less steep
- happy about suggestions...

### PR DESCRIPTION
**Pull request by veni-vidi-vici-dormivi**
- [ ] Closes #459
 - [x] Tests added

### CODE REVIEW

File: tests/unit/test_harmonic_model.py [hunk 1]
```diff
@@ -52,23 +52,21 @@ def test_generate_fourier_series_np():
...
-    [np.repeat([-1, 1], 10 * 6), np.linspace(-1, 1, 10 * 12) * 10],
-)
-def test_fit_to_bic_np(coefficients, yearly_predictor):
+def test_fit_to_bic_np(coefficients):
     max_order = 6
-    months = np.tile(np.arange(1, 13), 10)
-
+    n_years = 100
+    months = np.tile(np.arange(1, 13), n_years)
+    yearly_predictor = yearly_predictor = trend_data_1D(
```
Comment (mathause) [inline tests/unit/test_harmonic_model.py:59]:
```suggestion
    yearly_predictor = trend_data_1D(
```

File: tests/unit/test_harmonic_model.py [hunk 2]
```diff
@@ -52,23 +52,21 @@ def test_generate_fourier_series_np():
         np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
     ],
 )
-@pytest.mark.parametrize(
-    "yearly_predictor",
-    [np.repeat([-1, 1], 10 * 6), np.linspace(-1, 1, 10 * 12) * 10],
-)
-def test_fit_to_bic_np(coefficients, yearly_predictor):
+def test_fit_to_bic_np(coefficients):
```
Comment (mathause) [inline tests/unit/test_harmonic_model.py:55]:
```suggestion
def test_fit_to_bic_np(coefficients):
    # ensure original coeffs and ts is estimated from fourier series without noise
```

File: tests/unit/test_harmonic_model.py [hunk 3]
```diff
@@ -239,15 +145,34 @@ def test_fit_to_bic_xr():
...
     rng = np.random.default_rng(0)
     noisy_monthly_target = monthly_target + rng.normal(
         loc=0, scale=0.1, size=monthly_target.values.shape
     )
     result = fit_to_bic_xr(yearly_predictor, noisy_monthly_target)
-    xr.testing.assert_allclose(result["predictions"], monthly_target, atol=0.2)
+    xr.testing.assert_allclose(result["predictions"], monthly_target, atol=0.1)
+    np.testing.assert_allclose(
+        result.predictions.isel(cells=0, time=slice(0, 12)).values,
+        np.array(
```
Comment (mathause) [inline tests/unit/test_harmonic_model.py:159]: Can you move this out of the function?

**Review by mathause — APPROVED** [review]
Do you know if the tests are faster because they abort earlier or because the fitting is genuinely faster?

**Comment by veni-vidi-vici-dormivi** [comment]
> Do you know if the tests are faster because they abort earlier or because the fitting is genuinely faster?

Should be because they are generally faster because also before they aborted at low orders (for few time steps).

File: tests/unit/test_harmonic_model.py (hunk 2 above)
Comment (veni-vidi-vici-dormivi) [reply tests/unit/test_harmonic_model.py:55]: Do you really mean time steps here? I would say we try for the order?

File: tests/unit/test_harmonic_model.py (hunk 3 above)
Comment (veni-vidi-vici-dormivi) [reply tests/unit/test_harmonic_model.py:159]: What do you mean? Write a new one, e.g. `test_fit_to_bic_xr_withnoise` ?

**Comment by mathause** [comment]
Yes you are right it seems to be both :champagne:

File: tests/unit/test_harmonic_model.py (hunk 2 above)
Comment (mathause) [reply tests/unit/test_harmonic_model.py:55]: no ts -> timeseries; but yes that's not obvious - I updated the comment

File: tests/unit/test_harmonic_model.py (hunk 3 above)
Comment (mathause) [reply tests/unit/test_harmonic_model.py:159]: No, that you define a the array outside the `assert_allclose` function, e.g. `expected_predictions = np.array([...])`

Comment (veni-vidi-vici-dormivi) [reply tests/unit/test_harmonic_model.py:159]: Of course, thanks 😅

**Comment by mathause** [comment]
This brings down the time of the tests by 30 s (1/3) - another nice win :1st_place_medal:

**Pull request merged**